orbit/pkg/useraction/mdm_migration_darwin.go

```diff
@@ -563,3 +581,223 @@ func (m *swiftDialogMDMMigrator) MigrationInProgress() (bool, error) {
 	return m.mrw.RemoveFile()
 }
+
+func StartMDMMigrationOfflineWatcher(ctx context.Context, client *service.DeviceClient, swiftDialogPath string, swiftDialogCh chan struct{}, fileWatcher migration.FileWatcher) {
+	if cap(swiftDialogCh) != 1 {
+		log.Fatal().Msg("swift dialog channel must have a buffer size of 1")
+	}
+
+	watcher := &offlineWatcher{
+		client:          client,
+		swiftDialogPath: swiftDialogPath,
+		swiftDialogCh:   swiftDialogCh,
+		fileWatcher:     fileWatcher,
+	}
+
+	// start loop with 3-minute interval to ping server and show dialog if offline
+	go func() {
+		ticker := time.NewTicker(1 * time.Minute)
+		defer ticker.Stop()
+
+		log.Info().Msg("starting offline dialog loop")
+		for {
+			select {
+			case <-ctx.Done():
+				log.Info().Msg("stopping offline dialog loop")
+				return
+			case <-ticker.C:
+				log.Debug().Msg("offline dialog, got tick")
+				go watcher.processTick(ctx)
+			}
+		}
+	}()
+}
+
+type offlineWatcher struct {
+	client          *service.DeviceClient
+	swiftDialogPath string
+	// swiftDialogCh is shared with the migrator and used to ensure only one dialog is open at a time
+	swiftDialogCh chan struct{}
+	fileWatcher   migration.FileWatcher
+}
+
+func (o *offlineWatcher) processTick(ctx context.Context) {
+	// try to block the dialog channel
+	select {
+	case o.swiftDialogCh <- struct{}{}:
+		log.Debug().Msg("offline dialog, blocking dialog channel")
+	default:
+		log.Debug().Msg("offline dialog, dialog channel already blocked")
+		return
+	}
+
+	defer func() {
+		// non-blocking release of dialog channel
+		select {
+		case <-o.swiftDialogCh:
+			log.Debug().Msg("offline dialog, releasing dialog channel")
+		default:
+			// TODO: think through how this could happen in relation to the other processes using the dialog channel
+			log.Debug().Msg("offline dialog, dialog channel already released")
+		}
+	}()
+
+	if !o.isUnmanaged() || !o.isOffline() {
+		return
+	}
+
+	log.Info().Msg("showing offline dialog")
+	if err := o.showSwiftDialogMDMMigrationOffline(ctx); err != nil {
+		log.Error().Err(err).Msg("error showing offline dialog")
+	} else {
+		log.Info().Msg("done showing offline dialog")
+	}
+}
+
+func (o *offlineWatcher) isUnmanaged() bool {
+	mt, err := o.fileWatcher.GetMigrationType()
+	if err != nil {
+		// TODO: confirm error handling with jahziel
+		log.Error().Err(err).Msg("getting migration type")
+	}
+
+	if mt == "" {
+		// TODO: confirm with jahziel that this is the intended behavior for this case
+		log.Debug().Msg("offline dialog, no migration type found, do nothing")
+		return false
+	}
+
+	log.Debug().Msgf("offline dialog, device is unmanaged, migration type %s", mt)
+
+	// TODO: Maybe check show profiles and skip showing the dialog if the device is managed?
+
+	return true
+}
+
+func (o *offlineWatcher) isOffline() bool {
+	// TODO: should we rely on the Fleet server or should we use something else (e.g.,
+	// DNS lookup)?
```

Comment:
I think Fleet server is what we want, if you can't talk to Fleet, you can't enroll